rfcs/text/0019_lifecycle_preboot.md

```diff
@@ -0,0 +1,246 @@
+  readonly holdBootUntilResolved: (
+    promise: Promise<{ shouldReloadConfig: boolean } | undefined>
+  ) => void;
+}
+```
+
+Internal Preboot service contract will also expose `waitUntilCanBoot` method that bootstrap process can use to know when it can proceed to `setup` and `start` stages:
+
+```ts
+// NOTE(azasypkin): some existing properties have been omitted for brevity.
+export interface InternalPrebootServiceSetup {
+  readonly waitUntilCanBoot: () => Promise<{ shouldReloadConfig: boolean }>;
+}
+```
+
+### 3.2.10 Bootstrap
+
+We'll update Kibana bootstrap sequence to include `preboot` stage and to conditionally reload configuration before proceeding to `setup` and `start` stages:
+
+```ts
+// NOTE(azasypkin): some functionality and checks have been omitted for brevity.
+const { preboot } = await root.preboot();
+
+const { shouldReloadConfig } = await preboot.waitUntilCanBoot();
+if (shouldReloadConfig) {
+  await reloadConfiguration('pre-boot request');
+}
+
+await root.setup();
+await root.start();
+```
+
+It's not yet clear if we need to adjust the base path proxy to account for this new lifecycle stage (see [unresolved question 3](#83-development-mode-and-basepath-proxy)).
+
+# 4. Drawbacks
+
+The main drawback is that proposed changes affect quite a few Kibana Core services that may impose a risk of breaking something in the critical parts of Kibana.
+
+# 5. Alternatives
+
+The most viable alternative to support interactive setup mode for Kibana was a standalone application that would be completely separated from Kibana. We ruled out this option since we won't be able to leverage existing and battle-tested Core services, UI components, and development tools. This would make the long-term maintenance burden unreasonably high.
+
+# 6. Adoption strategy
+
+The new `preboot` stage doesn't need an adoption strategy since it's intended for internal platform use only.
+
+# 7. How we teach this
+
+The new `preboot` stage shouldn't need much knowledge sharing since it's intended for internal platform use only and doesn't affect the primary plugins. All new services, methods, and contracts will be sufficiently documented in the code.
+
+# 8. Unresolved questions
+
+## 8.1 Lifecycle stage name
+
+Is `preboot` the right name for this new lifecycle stage? Do we have a better alternative?
+
+## 8.2 Core client-side changes
+
+The server-side part of the `preboot` plugins will follow a new `PrebootPlugin` interface that doesn't have a `start` lifecycle stage, but the client-side part will stay the same as for primary plugins. This significantly simplifies implementation and doesn't introduce any known technical issues, but, unfortunately, brings some inconsistency to the codebase. Namely, the client-side `CoreStart` contract, that `preboot` plugins will have access to, will still expose methods to work with the services that won't be available at the `preboot` stage (e.g., Saved Objects service). Does this tradeoff sound reasonable?
```

Comment:
FWIW, when we discussed it with @azasypkin, I did agree that the pragmatic approach looked acceptable, at least initially.

Reply:
I agree with the approach to keep `setup` and `start` on the client. In the browser, `setup` and `start` occur once in every user window that is opened, so the semantics here are slightly different than the server. Keeping this consistent makes sense.

That said, I think the only client-side `CoreStart` services that need to unavailable are the `CoreStart.savedObjects` and `CoreStart.uiSettings` services. Could we expose a `PrebootPlugin` interface on the client that is identical but `Omit`s these two services from `PrebootCoreStart`?

Reply:
> That said, I think the only client-side `CoreStart` services that need to unavailable are the `CoreStart.savedObjects` and `CoreStart.uiSettings` services. Could we expose a `PrebootPlugin` interface on the client that is identical but `Omit`s these two services from `PrebootCoreStart`?

Yeah, we can definitely do that. Would "compile-time guarantee" we'll get with the dedicated client-side `PrebootPlugin` interface be enough or you think we should "physically" remove these two services from the start contract (I tried to avoid such client-side changes until now, but can investigate how difficult it would be if necessary)?